library/src/main/kotlin/build/buf/connect/impl/ProtocolClient.kt

```diff
@@ -200,4 +201,10 @@ class ProtocolClient(
             )
         )
     }
+
+    private fun <Input : Any, Output : Any> urlFromMethodSpec(methodSpec: MethodSpec<Input, Output>): URL {
+        val host = URI(config.host)
```

Comment:
Another point I raised in #12 was that the validation only happens when the `unary` or `stream` methods are called, instead of when the configuration object is originally constructed. I think that this makes it harder for a developer to discover that they've provided an invalid URL because they don't find out immediately.

For example, what happens if you do this?

```
fun createConnectClient(): ProtocolClientInterface {
    return ProtocolClient(
        httpClient = ConnectOkHttpClient(),
        ProtocolClientConfig(
            // We cannot use `server.toString()` because it adds a trailing slash
            host = "htp://example.com",
            serializationStrategy = GoogleJavaLiteProtobufStrategy(),
            networkProtocol = NetworkProtocol.CONNECT
        ),
    )
}
```

Note the `scheme` is `htp` instead of `http`. With the code as shown here, you only discover the typo later in the execution when you try to call the endpoint.

I think it would be helpful for the developer experience to know right away that an invalid value has been provided by doing this validation in the creation of the `config` object. Something like:

```
val baseUrl = URI(host)
```

Then you can use it here in `urlFromMethodSpec` with `config.baseUrl`.

Reply:
`URI()` doesn't actually do scheme validation. 

What I can do is to have a validation step on the configuration construction (which is basically `baseUri.toURL()`). If this is something that you're interested in for DX, I would like to separate that out in another PR and get connect-swift to also be consistent in behavior.

Reply:
I was thinking about okhttp's `toHttpUrl`, which I think does validation.

Also, I gave a bad example by giving a typo in the scheme as what we'd want validation for. A better example might be something that is completely malformed, which I think will fail creating a URI.

In any case, yes that can be a separate PR.

Thank you for looking at this.

Reply:
Oh yeah, this layer doesn't have the okhttp `toHttpUrl` dependency. We gotta use vanilla URI/URL